src/lib/temple/front/gas-station.ts

```diff
@@ -0,0 +1,199 @@
+import {
+  ContractAbstraction,
+  OpKind,
+  TezosToolkit,
+  Wallet,
+} from "@taquito/taquito";
+import { hex2buf } from "@taquito/utils";
+import BigNumber from "bignumber.js";
+
+import blake from "blakejs";
+import { FastRpcClient } from "lib/taquito-fast-rpc";
+import { michelEncoder } from "lib/temple/helpers";
+
+import { ReadOnlySigner } from "../read-only-signer";
+
+type ForgeTxParams = {
+  to: string;
+  tokenAddress: string;
+  tokenId?: number;
+  amount: BigNumber;
+  relayerFee: BigNumber;
+};
+
+type PermitParams = {
+  signature: string;
+  hash: string;
+  pubkey: string;
+  contractAddress: string;
+  callParams: any;
+};
+
+export const RELAYER_ADDRESS = "tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb";
+const ESTIMATOR_RPC_ENDPOINT = "http://167.99.133.200:8732";
+const ESTIMATOR_ADDRESS = "tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb";
+const ESTIMATOR_PUBLIC_KEY =
+  "edpkvGfYw3LyB1UcCahKQk4rF2tvbMUk8GFiTuMjL75uGXrpvKXhjn";
+
+const formTransferParams = (
+  from_: string,
+  relayerAddress: string,
+  { to: to_, tokenId, amount, relayerFee }: ForgeTxParams
+) => {
+  const intendedTx = { to_: to_, token_id: tokenId, amount: amount };
+
+  const feeTx = {
+    to_: relayerAddress,
+    token_id: tokenId,
+    amount: relayerFee,
+  };
+
+  const txList = [
+    [
+      {
+        from_: from_,
+        txs: [intendedTx, feeTx],
+      },
+    ],
+  ];
+
+  return txList;
+};
+
+const permitParamHash = async (
+  tezos: TezosToolkit,
+  contract: ContractAbstraction<Wallet>,
+  entrypoint: string,
+  parameters: any
+): Promise<string> => {
+  const raw_packed = await tezos.rpc.packData({
+    data: contract.parameterSchema.Encode(entrypoint, ...parameters),
+    // @ts-ignore
+    type: contract.parameterSchema.root.typeWithoutAnnotations(),
+  });
+
+  return blake.blake2bHex(hex2buf(raw_packed.packed), null, 32);
+};
+
+const getUnpackedUniques = (
+  contractAddress: string,
+  chainId: string,
+  currentPermitCount: number,
+  permitHash: string
+) => {
+  return {
+    data: {
+      prim: "Pair",
+      args: [
+        {
+          prim: "Pair",
+          args: [{ string: contractAddress }, { string: chainId }],
+        },
+        {
+          prim: "Pair",
+          args: [{ int: currentPermitCount.toString() }, { bytes: permitHash }],
+        },
+      ],
+    },
+    type: {
+      prim: "pair",
+      args: [
+        { prim: "pair", args: [{ prim: "address" }, { prim: "chain_id" }] },
+        { prim: "pair", args: [{ prim: "nat" }, { prim: "bytes" }] },
+      ],
+    },
+  };
+};
+
+const createPermitPayload = async (
+  tezos: TezosToolkit,
+  contractAddress: string,
+  entrypoint: string,
+  params: any
+) => {
+  const contract = await tezos.wallet.at(contractAddress);
+  const storage = await contract.storage<any>();
+  const signerKey = await tezos.signer.publicKey();
+  const paramHash = await permitParamHash(tezos, contract, entrypoint, params);
+
+  const chainId = await tezos.rpc.getChainId();
```

Comment:
Can we somehow pass known chain id from the wallet here?
Just to reduce the amount of requests to be performed

Reply:
it is already optimized under the hood.

Reply:
https://github.com/madfish-solutions/templewallet-extension/blob/develop/src/lib/taquito-fast-rpc.ts#L167